addons/binding/org.openhab.binding.neeo/src/main/java/org/openhab/binding/neeo/internal/discovery/NeeoRoomDiscoveryService.java

```diff
@@ -0,0 +1,237 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.neeo.internal.discovery;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.apache.commons.lang.StringUtils;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.common.ThreadPoolManager;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingRegistry;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.openhab.binding.neeo.NeeoConstants;
+import org.openhab.binding.neeo.NeeoUtil;
+import org.openhab.binding.neeo.handler.NeeoBrainHandler;
+import org.openhab.binding.neeo.internal.NeeoBrainApi;
+import org.openhab.binding.neeo.internal.NeeoBrainConfig;
+import org.openhab.binding.neeo.internal.models.NeeoBrain;
+import org.openhab.binding.neeo.internal.models.NeeoRoom;
+import org.openhab.binding.neeo.internal.type.NeeoTypeGenerator;
+import org.openhab.binding.neeo.internal.type.UidUtils;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of {@link AbstractDiscoveryService} that will discover the rooms in a NEEO brain;
+ *
+ * @author Tim Roberts - Initial contribution
+ */
+@Component(service = DiscoveryService.class, immediate = true)
+public class NeeoRoomDiscoveryService extends AbstractDiscoveryService {
+    /** The logger */
+    private final Logger logger = LoggerFactory.getLogger(NeeoRoomDiscoveryService.class);
+
+    /** The scanning task (not-null when connecting, null otherwise) */
+    private final AtomicReference<@Nullable Future<?>> scan = new AtomicReference<>();
+
+    /** The scheduler used to schedule tasks */
+    private final ScheduledExecutorService scheduler = ThreadPoolManager.getScheduledPool(NeeoConstants.THREADPOOL_ID);
+
+    /** The thing registry that we use */
+    @NonNullByDefault({})
+    private ThingRegistry thingRegistry;
```

Comment:
A discovery service must not use the thingregistry, please remove.

Reply:
Let's table this and the other discovery comments to see what the end result of https://github.com/eclipse/smarthome/issues/5582 (which I'll be adding comments to eventually).  I really don't like the approach you described because it sprinkles the discovery around a few classes.

Reply:
Alright, and while you are positing to eclipse/smarthome#5582, please link this conversation.